### ISSUE TITLE
[UX] User edit form: conditionally show the "Current password" field

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3389]
Here's what this form currently looks like:

<img width="827" alt="screen shot 2018-11-21 at 2 57 50 pm" src="https://user-images.githubusercontent.com/2423362/48824890-c8e53f80-edb9-11e8-8405-930691562bea.png">

There's quite a few things that annoy me here:

- I would prefer the all password fields to be rendered together; one after the other, without any other field in between.
- That "Current password" field is only required if the user needs to change their email and/or their password, but it is shown **always**.
- The help text under the "New password" field explains that another field is required instead of explaining what the field it belongs to does.

Lets fix all these things 😄 

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2389

**Comment by klonos** [comment]
PR up for review...

- The "Current password" field is hidden by default:

  <img width="824" alt="screen shot 2018-11-21 at 6 14 25 pm" src="https://user-images.githubusercontent.com/2423362/48825472-950b1980-edbb-11e8-80fb-5200df97cccf.png">

- If the value of the email has been edited (a dot added in this screenshot), then the "Current password" field is shown:

  <img width="818" alt="screen shot 2018-11-21 at 6 14 38 pm" src="https://user-images.githubusercontent.com/2423362/48825548-cd125c80-edbb-11e8-845a-6a2ad39e7056.png">

- If any character is entered in the password field, then the "Current password" field is shown:

  <img width="824" alt="screen shot 2018-11-21 at 6 14 53 pm" src="https://user-images.githubusercontent.com/2423362/48825599-f16e3900-edbb-11e8-8af8-8d80ea419875.png">

- The help text for the "New password" field has been changed to:

  > Use this field to change your current password to a new one.

- A red "required" asterisk is added to the "Current password" field, but due to known issues with the Form API, `#states`, and the `'required'` state this does not do anything validation-wise, so `user_validate_current_pass()` is still required. See https://www.drupal.org/project/drupal/issues/2855139 https://www.drupal.org/node/2405271 https://www.drupal.org/project/conditional_fields/issues/1561272

**Comment by herbdool** [comment]
I don't think this is a bug, strictly speaking. It's a UX improvement. I'm also not convinced it should be changed in a patch release.

**Comment by klonos** [comment]
Thanks @herbdool ...I have to admit that "technically" this is not a bug per se ...or not an apparent one at least, but AFAIK historically we have been permitting fixes to UX issues to be committed in patch releases; whether these issues where tagged as bugs or mere tasks (UX improvements). I could be wrong, but that is my understanding. @jenlampton ??

Having said that, the actual bug around what is going on in this form UX-wise has been marked as a major bug in d.org: [#states required is not working in form API](https://www.drupal.org/project/drupal/issues/2855139). This issue here is not trying to address the underlying bug with the Form API, but it does prevent the following UX WTF (pretend that you are not a seasoned Drupal/Backdrop user for a moment 😄):

1. Edit your user account and change your email address
2. Since no other field is marked as required in that form (and you don't speak Drupal/Backdrop, remember?), scroll down the page and click "Save" -> a required, but also non-required field ...WFT?!?

...hence the assignment of the bug label for this issue.

**Comment by klonos** [comment]
I brought this up during the design meeting, and the general consensus seemed to be that we should have tests for this. Pushing to 1.12.0 because I don't speak tests, and I have also added the "help wanted" tag for anyone for whom tests are their thing 😅.

**Comment by herbdool** [comment]
@klonos that's what I understand too: that small UX improvements can be in a minor release. However, I don't think they need to be tagged as bugs. It helps when only things that are actually broken are tagged as bugs. This helps us prioritize.

**Comment by jenlampton** [comment]
We also mentioned in that meeting that we don't have any way to do testing in JavaScript - so we'll need to determine what testing we can do on the PHP side, and see if those tests aren't already being done. 

I have thoughts about the text in the PR:

> Use this field to change your current password to a new one.

Does this add value beyond the "New password" label? IMHO it doesn't and we should remove the description text here entirely. 

> Required if you want to change the email address or specify a new password. If you have forgotten your current password, you can reset it via email.

How about ...

`Required to change your email address or password. If you have forgotten your current password you must reset it by email.`

But we should be careful to only add `If you have forgotten your current password you must reset it by email.` if you are looking at your own page. Will this text appear for admins?

**Comment by docwilmot** [comment]
I like this issue. Ideas:
I think both password fields should be hidden; there should be a "Change password" link that toggles the new and current password fields to appear. 
And use JS to toggle the current password field when JS detects youve typed a full valid email into the email field, not just a dot.

**Comment by jenlampton** [comment]
> And use JS to toggle the current password field when JS detects youve typed a full valid email into the email field, not just a dot.

I kinda like that the password fields when any change is made. It's very clear this way that changing the email requires a password confirmation.

**Comment by klonos** [comment]
Thanks for the reviews @jenlampton and @docwilmot 👍

...I have used `#states` for the hiding/unhiding magic on this form @docwilmot, and my JS skills are still laughable. So if we were to implement a non-`#states` JS toggle here I might have to step back and leave this to someone else.

@jenlampton I would not mind filing a second, alternative PR that implements a JS toggle. This would give any reviewers the chance to compare the UX with both implementations, and then compare as to which one is the most painless.

...unless I cheat and use a checkbox styled as button (like https://stackoverflow.com/questions/7642277/css-styled-a-checkbox-to-look-like-a-button-is-there-a-hover for example). That would allow me to still use `#states` under the hood 😈. Would such a solution be acceptable?

**Comment by docwilmot** [comment]
> I kinda like that the password field shows up when any change is made. It's very clear this way that changing the email requires a password confirmation.

Its also annoying when fields start popping in and out of existence while your interacting with one area of the UI; you get a what-the-hell-just-happened feeling when you sense theres a change before you figure out "oh I think a new password box just appeared". Better I think, when youre done, or you remove focus, to then guide you to the next step, ideally with a message .

**Comment by jenlampton** [comment]
Bumping to 1.13 milestone.

**Comment by jenlampton** [comment]
oops, sorry for closing @klonos. Why is there a close and comment button anyway!

**Comment by quicksketch** [comment]
Agreed this probably is too much for a bugfix release, so this may have to wait for 1.13, unless we come up with something that is very minor and low-impact.

I'm not a fan of the checkbox. The "checkbox shows another field" pattern is common, but I've not seen it used on an account page like this. The combination of the checkbox OR changing email address is also a strange one for me.

Something that I'm seeing more commonly is an entirely separate page for changing sensitive information like password or email address. Here on GitHub or on Facebook you have to enter your password before you can even access the form to change your password. Perhaps we could do something like that? Some more options:

- Block access to the entire /user/x/edit form until you enter your current password.
- Require your current password _after_ submitting the form, in a whole-page display. Then after entering your password save the form.
- Show fields for email and password but disabled, but link to a completely new page, like `/user/x/edit-password` or `/user/x/edit-email`, where a password is required to change those values.

**Comment by jenlampton** [comment]
No attention here since Jan 15. Pulling milestone, adding candidate label.

**Comment by klonos** [comment]
Things changed quite a bit after #4266 so this will need to be reworked.

I've noticed a bug as well, after the 1.16.0 release. Here's what the user edit form looks like now:

![Screen Shot 2020-05-19 at 2 03 44 am](https://user-images.githubusercontent.com/2423362/82267455-13c32180-9975-11ea-84c3-fe183f423a4c.png)

The help text "Current password must be entered to set a new password." does not make sense under the "New password" field, and in fact there is no validation - the password seems to be changed w/o requesting the old/current password.

**Comment by BWPanda** [comment]
> Require your current password after submitting the form, in a whole-page display. Then after entering your password save the form.

While reading through this issue I literally had the same thought. However I was thinking of the "current password" field showing in a modal instead.

So workflow would be:
1. Go to user account form
2. See email and new password fields
3. Edit one or both of them
4. Press Save
5. Modal pops up with current password field (and possibly message explaining this is for authorisation)
6. Enter current/old password
7. Save

And obviously the same is true if we decide to implement this as a separate page instead of a modal. And if the current password form is cancelled, go back to user account form with old values (I.e. nothing saved).
